src/fs.rs

```diff
@@ -97,7 +112,217 @@ fn write_directory_contents(dir: &Directory, fd: &OwnedFd, repo: &Repository) ->
     Ok(())
 }
 
+// NB: hardlinks not supported
 pub fn write_to_path(repo: &Repository, dir: &Directory, output_dir: &Path) -> Result<()> {
     let fd = openat(CWD, output_dir, OFlags::PATH | OFlags::DIRECTORY, 0.into())?;
     write_directory_contents(dir, &fd, repo)
 }
+
+pub struct FilesystemReader<'repo> {
+    st_dev: u64,
+    repo: Option<&'repo Repository>,
+    inodes: HashMap<u64, Rc<Leaf>>,
+    root_mtime: i64,
+}
+
+impl<'repo> FilesystemReader<'repo> {
+    fn read_xattrs(&mut self, fd: &OwnedFd) -> Result<BTreeMap<Box<OsStr>, Box<[u8]>>> {
+        // flistxattr() and fgetxattr() don't with with O_PATH fds, so go via /proc/self/fd. Note:
+        // we want the symlink-following version of this call, which produces the correct behaviour
+        // even when trying to read xattrs from symlinks themselves.  See
+        // https://gist.github.com/allisonkarlitskaya/7a80f2ebb3314d80f45c653a1ba0e398
+        let filename = proc_self_fd(fd);
+
+        let mut xattrs = BTreeMap::new();
+
+        let names_size = listxattr(&filename, &mut [])?;
+        let mut names = vec![0; names_size];
+        let actual_names_size = listxattr(&filename, &mut names)?;
+        ensure!(
+            actual_names_size == names.len(),
+            "xattrs changed during read"
+        );
+
+        let names: Vec<u8> = names.into_iter().map(|c| c as u8).collect(); // fml
+
+        let mut buffer = [0; 65536];
+        for name in names.split_inclusive(|c| *c == 0) {
+            let name = CStr::from_bytes_with_nul(name)?;
+            let value_size = getxattr(&filename, name, &mut buffer)?;
+            let key = Box::from(OsStr::from_bytes(name.to_bytes()));
+            let value = Box::from(&buffer[..value_size]);
+            xattrs.insert(key, value);
+        }
+
+        Ok(xattrs)
+    }
+
+    fn stat(&mut self, fd: &OwnedFd, ifmt: FileType) -> Result<(rustix::fs::Stat, Stat)> {
+        let buf = fstat(fd)?;
+
+        ensure!(
+            FileType::from_raw_mode(buf.st_mode) == ifmt,
+            "File type changed
+            between readdir() and fstat()"
+        );
+
+        let mtime = buf.st_mtime as i64;
+
+        if buf.st_dev != self.st_dev {
+            if self.st_dev == u64::MAX {
+                self.st_dev = buf.st_dev;
+            } else {
+                bail!("Attempting to cross devices while importing filesystem");
```

Comment:
There's also https://docs.rs/openat2/latest/openat2/struct.ResolveFlags.html#associatedconstant.NO_XDEV

Reply:
This is really interesting, and I didn't know about it.  Thanks.

rustix binds that via https://doc.servo.org/rustix/backend/fs/types/struct.ResolveFlags.html

The trouble with this is that it would make reading the root directory be different from all the other inodes: entering the bindmount from `CWD` would be considered a violation of the `NO_XDEV` rule.  Still, though: we have a bunch of other ugly code here that's ugly specifically because the root node *isn't* handled separately, and it ought to be.

I'm going to try to see if I can use this flag and unroll the root directory case a bit to get something a bit nicer all around.  Thanks!